dot_config/nvim/lua/_lsp.lua

```diff
@@ -0,0 +1,249 @@
+  vim.api.nvim_command('autocmd CursorHold <buffer> lua require("lspsaga.diagnostic").show_line_diagnostics()')
+
+  vim.api.nvim_command('highlight LspDiagnosticsDefaultError guifg=#BF616A')
+  vim.api.nvim_command('highlight LspDiagnosticsDefaultWarning guifg=#EBCB8B')
+  vim.api.nvim_command('highlight LspDiagnosticsDefaultInformation guifg=none')
+  vim.api.nvim_command('highlight LspDiagnosticsDefaultHint guifg=none')
+
+  vim.api.nvim_command('highlight LspDiagnosticsUnderlineError guifg=#BF616A gui=underline')
+  vim.api.nvim_command('highlight LspDiagnosticsUnderlineWarning guifg=#EBCB8B gui=underline')
+  vim.api.nvim_command('highlight LspDiagnosticsUnderlineInformation guifg=none gui=underline')
+  vim.api.nvim_command('highlight LspDiagnosticsUnderlineHint guifg=none gui=underline')
+
+  vim.fn.sign_define("LspDiagnosticsSignError", { text = "⨯", texthl = "LspDiagnosticsSignError" })
+  vim.fn.sign_define("LspDiagnosticsSignWarning", { text = "◆", texthl = "LspDiagnosticsSignWarning" })
+  vim.fn.sign_define("LspDiagnosticsSignInformation", { text = "ⓘ ", texthl = "LspDiagnosticsSignInformation" })
+  vim.fn.sign_define("LspDiagnosticsSignHint", { text = "…", texthl = "LspDiagnosticsSignHint" })
+
+  -- Mappings.
+  local opts = { noremap=true, silent=true }
+  buf_set_keymap('n', 'gD',         '<cmd>lua vim.lsp.buf.definition()<CR>', opts)
+  buf_set_keymap('n', 'gd',         '<cmd>Lspsaga preview_definition<CR>', opts)
+  buf_set_keymap('n', 'gi',         '<cmd>lua vim.lsp.buf.implementation()<CR>', opts)
+  buf_set_keymap('n', 'gr',         '<cmd>Telescope lsp_references<CR>', opts)
+  buf_set_keymap('n', 'K',          '<cmd>Lspsaga hover_doc<CR>', opts)
+  buf_set_keymap('n', '<leader>ac', '<cmd>Telescope lsp_code_actions<CR>', opts)
+
+  -- other mappings, not sure about these
+  buf_set_keymap('n', '<space>wa',  '<cmd>lua vim.lsp.buf.add_workspace_folder()<CR>', opts)
+  buf_set_keymap('n', '<space>wr',  '<cmd>lua vim.lsp.buf.remove_workspace_folder()<CR>', opts)
+  buf_set_keymap('n', '<space>wl',  '<cmd>lua print(vim.inspect(vim.lsp.buf.list_workspace_folders()))<CR>', opts)
+  buf_set_keymap('n', '<space>D',   '<cmd>lua vim.lsp.buf.type_definition()<CR>', opts)
+  buf_set_keymap('n', '<space>rn',  '<cmd>lua vim.lsp.buf.rename()<CR>', opts)
+  buf_set_keymap('n', '<space>e',   '<cmd>lua vim.lsp.diagnostic.show_line_diagnostics()<CR>', opts)
+  buf_set_keymap('n', '[d',         '<cmd>lua vim.lsp.diagnostic.goto_prev()<CR>', opts)
+  buf_set_keymap('n', ']d',         '<cmd>lua vim.lsp.diagnostic.goto_next()<CR>', opts)
+  buf_set_keymap('n', '<space>q',   '<cmd>lua vim.lsp.diagnostic.set_loclist()<CR>', opts)
+
+  -- Set some keybinds conditional on server capabilities
+  if client.resolved_capabilities.document_formatting then
+    vim.cmd [[augroup lsp_formatting]]
+    vim.cmd [[autocmd!]]
+    vim.cmd [[autocmd BufWritePre <buffer> :lua vim.lsp.buf.formatting_sync({}, 5000)]]
```

Comment:
Had to keep this (`5000`) high to ensure it always writes - don't want to have double writes as will be annoying for things like jest, webpack, etc.